File: common/src/main/java/com/pluxity/cctv/CctvService.java
package com.pluxity.cctv;

import com.pluxity.cctv.category.CctvCategory;
import com.pluxity.cctv.category.CctvCategoryService;
import com.pluxity.cctv.category.dto.CctvCategoryResponse;
import com.pluxity.cctv.dto.CctvCreateRequest;
import com.pluxity.cctv.dto.CctvResponse;
import com.pluxity.cctv.dto.CctvUpdateRequest;
import com.pluxity.feature.dto.FeatureResponse;
import com.pluxity.feature.entity.Feature;
import com.pluxity.feature.entity.FeatureType;
import com.pluxity.feature.service.FeatureFacade;
import com.pluxity.global.constant.ErrorCode;
import com.pluxity.global.exception.CustomException;
import com.pluxity.global.utils.MappingUtils;
import jakarta.validation.Valid;
import java.util.List;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

@Service
@RequiredArgsConstructor
public class CctvService {

    private final CctvRepository cctvRepository;
    private final CctvCategoryService cctvCategoryService;
    private final FeatureFacade featureFacade;

    @Transactional
    public String create(@Valid CctvCreateRequest request) {
        CctvCategory category =
                MappingUtils.findByIdIfExists(request.categoryId(), cctvCategoryService::findById);
        return cctvRepository
                .save(
                        Cctv.builder()
                                .id(request.id())
                                .name(request.name())
                                .url(request.url())
                                .category(category)
                                .build())
                .getId();
    }

    @Transactional(readOnly = true)
    public List<CctvResponse> findAll() {
        List<Cctv> list = cctvRepository.findAll();
        return list.stream().map(this::createResponse).toList();
    }

    @Transactional(readOnly = true)
    public CctvResponse getById(String id) {
        return createResponse(findById(id));
    }

    @Transactional(readOnly = true)
    public CctvResponse getByFeatureId(String featureId) {
        Feature feature = featureFacade.findById(featureId);
        Cctv cctv =
                cctvRepository
                        .findByFeature(feature)
                        .orElseThrow(() -> new CustomException(ErrorCode.NOT_FOUND_CCTV_BY_FEATURE, featureId));
        return createResponse(cctv);
    }

    private CctvResponse createResponse(Cctv cctv) {
        return CctvResponse.builder()
                .id(cctv.getId())
                .name(cctv.getName())
                .url(cctv.getUrl())
                .feature(cctv.getFeature() != null ? FeatureResponse.from(cctv.getFeature()) : null)
                .cctvCategory(
                        cctv.getCategory() != null ? CctvCategoryResponse.from(cctv.getCategory()) : null)
                .build();
    }

    @Transactional
    public void update(String id, CctvUpdateRequest request) {
        Cctv cctv = findById(id);
        cctv.updateCctv(request);
        CctvCategory category = cctvCategoryService.findById(request.categoryId());
        cctv.changeCategory(category);
    }

    @Transactional
    public void delete(String id) {
        Cctv cctv = findById(id);
        Feature feature = cctv.getFeature();
        feature.updateFeatureType(FeatureType.NONE);
        cctvRepository.delete(cctv);
    }

    @Transactional(readOnly = true)
    public Cctv findById(String id) {
        return cctvRepository
                .findById(id)
                .orElseThrow(() -> new CustomException(ErrorCode.NOT_FOUND_CCTV, id));
    }
}

File: common/src/main/java/com/pluxity/cctv/CctvService.kt
package com.pluxity.cctv

import com.pluxity.cctv.dto.CctvCreateRequest
import com.pluxity.cctv.dto.CctvResponse
import com.pluxity.cctv.dto.CctvUpdateRequest
import com.pluxity.cctv.dto.toCctvResponse
import com.pluxity.cctv.entity.Cctv
import com.pluxity.cctv.repository.CctvRepository
import com.pluxity.cctv.repository.DeviceCctvRepository
import com.pluxity.global.constant.ErrorCode
import com.pluxity.global.exception.CustomException
import org.springframework.data.domain.Sort
import org.springframework.data.repository.findByIdOrNull
import org.springframework.stereotype.Service
import org.springframework.transaction.annotation.Transactional

@Service
class CctvService(
    private val cctvRepository: CctvRepository,
    private val deviceCctvRepository: DeviceCctvRepository,
) {
    @Transactional
    fun create(request: CctvCreateRequest): String = cctvRepository.save(Cctv(id = request.id, name = request.name, url = request.url)).id

    @Transactional(readOnly = true)
    fun findAll(): List<CctvResponse> {
        val list = cctvRepository.findAll(Sort.by(Sort.Direction.DESC, "createdAt"))
        return list.map { it.toCctvResponse() }
    }

    @Transactional(readOnly = true)
    fun getById(id: String): CctvResponse {
        val cctv = findById(id)
        return cctv.toCctvResponse()
    }

    @Transactional
    fun update(
        id: String,
        request: CctvUpdateRequest,
    ) {
        val cctv = findById(id)
        cctv.updateCctv(request)
    }

    @Transactional
    fun delete(id: String) {
        val cctv = findById(id)
        deviceCctvRepository.deleteByCctvIdIn(listOf(cctv.id))
        cctvRepository.deleteById(cctv.id)
    }

    @Transactional(readOnly = true)
    fun findById(id: String): Cctv =
        cctvRepository.findByIdOrNull(id)
            ?: throw CustomException(ErrorCode.NOT_FOUND_CCTV, id)
}
